### PR TITLE
Huddles!

### PR DESCRIPTION
**Pull request by colinhowe**
This PR implements:

- Proper video huddles
- An experiment to allow for a video stream to be focused. This will be tidied up when we do "talk show studio"
- Removal of custom buttons - it isn't needed and over complicates things

### CODE REVIEW

File: src/components/attendee/VideoHuddle/VideoHuddle.module.scss [hunk 1]
```diff
@@ -19,24 +19,35 @@
...
   display: flex;
   flex-direction: column;
-  justify-self: flex-start;
+  justify-self: start;
   border-left: 1px solid theme.$sparkle--border;
+
+  span {
+    border-bottom: 1px solid theme.$sparkle--border;
+    width: 32px;
+    height: 32px;
```
Comment (colinhowe) [inline src/components/attendee/VideoHuddle/VideoHuddle.module.scss:38]: Temporary numbers until I talk over the details with Sabrina/Matt

File: src/components/templates/ExperimentalSpace/ExperimentalSpace.tsx [hunk 2]
```diff
@@ -0,0 +1,195 @@
+import React, { useCallback, useEffect, useMemo } from "react";
```
Comment (colinhowe) [inline src/components/templates/ExperimentalSpace/ExperimentalSpace.tsx:1]: This file doesn't need reviewing. It's a space type purely for prototyping "projecting" video to a central pane. It'll be deleted.

File: src/components/templates/Jazzbar/JazzBar/JazzBar.tsx [hunk 3]
```diff
@@ -86,10 +86,11 @@ export const JazzBar: React.FC<JazzProps> = ({ venue }) => {
 
   const joinTable = useCallback(
     (table) => {
-      joinHuddle(`${venue.id}-${table}`);
+      // @debt TOOD Fix this
```
Comment (colinhowe) [inline src/components/templates/Jazzbar/JazzBar/JazzBar.tsx:89]: There's a `jazzbar` branch that deals with this

File: src/components/attendee/VideoHuddle/VideoHuddle.module.scss (hunk 1 above)
Comment (goran-peoski-work) [reply src/components/attendee/VideoHuddle/VideoHuddle.module.scss:38]: Private constants to the module `$-size: 32px`, but also, this reminds me we didn't bring `square-size` mixin from the old design.

File: src/components/templates/ExperimentalSpace/ExperimentalSpace.tsx (hunk 2 above)
Comment (goran-peoski-work) [reply src/components/templates/ExperimentalSpace/ExperimentalSpace.tsx:1]: I guess the above comment should have been useful inside the file

File: src/components/templates/Jazzbar/JazzBar/JazzBar.tsx (hunk 3 above)
Comment (goran-peoski-work) [reply src/components/templates/Jazzbar/JazzBar/JazzBar.tsx:89]: Which one gets to be merged first and into which base?

**Review by goran-peoski-work — APPROVED** [review]
More questions than change requests. Some non-obligatory suggestions.

File: firestore.rules [hunk 4]
```diff
@@ -42,6 +42,10 @@ service cloud.firestore {
     match /experiences/{experienceId}/{restOfPath=**} {
       allow read, write: if request.auth.uid != null;
     }
+    // TODO This should not make it into production, it's an experiment
```
[the remark this answers is not in the document]
Comment (colinhowe) [reply firestore.rules:45]: I want to do a grep for TODO before we merge the redesign branch into staging. We shouldn't be introducing any new ones.

File: src/api/experiments.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import firebase from "firebase/compat/app";
+
+import { COLLECTION_EXPERIMENTS } from "settings";
+
+// This is a temporary API until we have something better. For now we use
+// a collection "huddles" to store all this temporary data until we have a
+// better place.
+
+export const setProjectedVideoTrackId = async (
```
[the remark this answers is not in the document]
Comment (colinhowe) [reply src/api/experiments.ts:9]: It's a track that is made larger and displayed in the middle of a space. Think of it like "this is the main speaker / screen share"

File: src/components/attendee/AttendeeLayout.tsx [hunk 6]
```diff
@@ -4,25 +4,27 @@ import { withFallback } from "components/hocs/gate/withFallback";
...
 import "scss/attendee/initial.scss";
 import styles from "./AttendeeLayout.module.scss";
 
-const _AttendeeLayout: React.FC = () => {
-  const { userId } = useUser();
+interface _AttendeeLayoutProps {
+  userId: string;
+}
 
+const _AttendeeLayout: React.FC<_AttendeeLayoutProps> = ({ userId }) => {
```
[the remark this answers is not in the document]
Comment (colinhowe) [reply src/components/attendee/AttendeeLayout.tsx:27]: The diff doesn't show the creation of AttendeeLayout below this block:

```
export const AttendeeLayout = compose(
  withAuth,
  withFallback(["isAuthLoaded", "userId"], NotLoggedInFallback),
  withRequired({
    required: ["userId"],
    fallback: LoadingPage,
  })
)(_AttendeeLayout);
```

File: src/components/attendee/VideoComms/VideoCommsParticipant.tsx [hunk 7]
```diff
@@ -38,64 +33,101 @@ export const VideoCommsParticipant: React.FC<VideoCommsParticipantProps> = ({
...
+              sourceType={VideoSource.Webcam}
+            />
+          )}
+        </div>
       </div>
+      {screenshareTrack?.enabled && (
+        <div className={styles.trackContainer}>
+          <VideoTrackDisplay track={screenshareTrack} />
+          {isLocal && (
+            <div className={determineControlsClassnames(true)}>
```
[the remark this answers is not in the document]
Comment (colinhowe) [reply src/components/attendee/VideoComms/VideoCommsParticipant.tsx:119]: Simplified this code with the suggested change to determineControlsClassnames

File: src/components/attendee/VideoComms/internal/TwilioClient.tsx [hunk 8]
```diff
@@ -283,6 +283,23 @@ export const TwilioClient = (onStateUpdateCallback: StateUpdateCallback) => {
...
+      return;
+    }
+    for (const track of room.localParticipant.videoTracks.values()) {
+      if (track.trackName !== TRACK_NAME_SCREENSHARE) {
+        // Only disable screenshare tracks
+        continue;
+      }
+      track.track.disable();
+      track.track.stop();
+      room.localParticipant.unpublishTrack(track.track);
```
[the remark this answers is not in the document]
Comment (colinhowe) [reply src/components/attendee/VideoComms/internal/TwilioClient.tsx:298]: Twilio has some bad naming :-/

File: src/components/attendee/VideoComms/internal/TwilioClient.tsx [hunk 9]
```diff
@@ -350,7 +367,11 @@ export const TwilioClient = (onStateUpdateCallback: StateUpdateCallback) => {
       console.warn("startVideo called from invalid state");
       return;
     }
-    if (!room.localParticipant.videoTracks.size) {
+    const webcamTracks = Array.from(
+      room.localParticipant.videoTracks.values()
+    ).filter(({ trackName }) => trackName !== TRACK_NAME_SCREENSHARE);
```
[the remark this answers is not in the document]
Comment (colinhowe) [reply src/components/attendee/VideoComms/internal/TwilioClient.tsx:372]: Agreed

File: src/components/attendee/VideoHuddle/HuddleProvider.tsx [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+export const HuddleContext = React.createContext<HuddleContextType>({
+  inHuddle: false,
+  setInHuddle: () => {},
+});
+
+interface HuddleProviderProps {
+  children: React.ReactNode;
+}
+
+export const HuddleProvider: React.FC<HuddleProviderProps> = ({ children }) => {
```
[the remark this answers is not in the document]
Comment (colinhowe) [reply src/components/attendee/VideoHuddle/HuddleProvider.tsx:17]: Fixed

File: src/components/attendee/VideoHuddle/VideoHuddle.module.scss (hunk 1 above)
Comment (colinhowe) [reply src/components/attendee/VideoHuddle/VideoHuddle.module.scss:38]: Good idea

**Pull request closed without merging**